azurelinuxagent/ga/update.py

```diff
@@ -88,6 +93,29 @@
 ]
 
 
+class ExtensionsSummary(object):
+    """
+    The extensions summary is a list of (extension name, extension status) tuples for the current goal state; it is
+    used to report changes in the status of extensions and to keep track of when the goal state converges (i.e. when
+    all extensions in the goal state reach a terminal state: success or error.)
+    The summary is computed from the VmStatus reported to blob storage.
+    """
+    def __init__(self, vm_status=None):
+        if vm_status is None:
+            self.summary = []
+            self.converged = True
+        else:
+            self.summary = [(h.extension_status.name, h.extension_status.status) for h in vm_status.vmAgent.extensionHandlers]
+            self.summary.sort(key=lambda s: s[0])  # sort by extension name to make comparisons easier
```

Comment:
This will throw if the name is None. 
It generally shouldn't be None but would be safer to have a check

Reply:
how can the extension name be None?

Reply:
i guess my point is that name is not special, why check for name not to be None, but not check status or vmAgent, or extensionHandlers?

rather, we should ensure we handle exceptions correctly if at some point a bug is introduced that makes any of those fields to be undefined. i realized the new UpdateHandler._report_status should handle errors in a similar way to ExtHandlersHandler.report_ext_handlers_status; I'll add that.